diesel/src/query_source/mod.rs

```diff
@@ -46,26 +46,31 @@ pub trait QuerySource {
 /// been generated by the [`table!` macro](crate::table!).
 pub trait Column: Expression {
     /// The table which this column belongs to
-    type Table: Table;
+    type Source: QuerySource;
 
     /// The name of this column
     const NAME: &'static str;
 }
 
+/// A SQL database view. Types which implement this trait should have been
+/// generated by the [`view!` macro](crate::view!).
+pub trait View: QuerySource + AsQuery + Sized {
+    /// The type returned by `all_columns`
+    type AllColumns: SelectableExpression<Self> + ValidGrouping<()>;
+    /// Returns a tuple of all columns belonging to this table.
+    fn all_columns() -> Self::AllColumns;
+}
+
 /// A SQL database table. Types which implement this trait should have been
 /// generated by the [`table!` macro](crate::table!).
-pub trait Table: QuerySource + AsQuery + Sized {
+pub trait Table: View {
```

Comment:
SQL-wise, a table is not a view, so I think we need a different naming for the concept of "either a view or a table", because the code is otherwise going to be very hard to read.
(This is linked to https://github.com/diesel-rs/diesel/pull/4077#discussion_r1651152998)

Reply:
Yes, I also agree there's something wrong here.

Reply:
(at least in Postgres) Views, Materialized Views and Tables [are all "Relations"](https://www.postgresql.org/docs/16/glossary.html#GLOSSARY-RELATION) so maybe that's a good starting point when it comes to naming?

Reply:
It mentions that "composite types and indexes are relations" as well but that may still be a good naming, if we need one that's separate from QuerySource.

Reply:
I agree with @Ten0 here that `View` shouldn't be the super trait of `Table`. Instead I would propose the following structure:

```rust
trait Relation: QuerySource + AsQuery + Sized {
    type AllColumns: SelectableExpression<Self> + ValidGrouping<()>;

    /// Returns a tuple of all columns belonging to this table.
    fn all_columns() -> Self::AllColumns;
}

trait Table: Relation {
    /// The type returned by `primary_key`
    type PrimaryKey: SelectableExpression<Self> + ValidGrouping<()>;

    /// Returns the primary key of this table.
    ///
    ///
    /// If the table has a composite primary key, this will be a tuple.
    /// If the table has a composite primary key, this will be a tuple.
    fn primary_key(&self) -> Self::PrimaryKey;
}

trait View: Relation {
 // for now only a marker trait?
}
```

I think the "Relation" name suggested by @czotomo is a good choice here for the new trait with the shared behavior. 

By having the super trait relation we also ensure that this isn't a breaking change and by having a seperate `View` trait that is not implemented for `Table` (and `Table` is not implemented for `View` by default) we can use that to differentiate behavior between these two. Also I think this opens up the road to add more table like constructions, like sql functions returning a "table`, in the future more easily.

For the time being I wouldn't want to expose the structures purely based on the `Relation` trait (without implementing `Table` nor `View`) in an easy way for the user.

Reply:
I had another look at this and unfortunately the suggested trait structure won't be possible without breaking changes. In fact we cannot remove items from the `Table` trait without causing locations that manually implement this trait to break.

I think we might be able to workaround this issue by using the following structure:

```rust
trait Relation: QuerySource + AsQuery + Sized {
    type AllColumns: SelectableExpression<Self> + ValidGrouping<()>;

    /// Returns a tuple of all columns belonging to this table.
    fn all_columns() -> Self::AllColumns;
}

trait Table: QuerySource {
    // unchanged
}

impl<T> Relation for T where T: Table {
    // forward relevant types/functions to the new trait
}

trait View: Relation {
 // for now only a marker trait?
}
```

And then change most of the existing `AsQuery` impls to use the new `Relation` trait instead. By that we won't break any existing usage, but can use the new trait in the relevant places. We might want to have a look at the impact on the compiler error messages there.